thewalrus/tests/test_csamples.py

```diff
@@ -65,3 +69,29 @@ def test_dist_thermal():
     rel_freq = freq / n_samples
     expected = (1 / (1 + n_mean)) * (n_mean / (1 + n_mean)) ** (np.arange(len(rel_freq)))
     assert np.allclose(rel_freq, expected, atol=10 / np.sqrt(n_samples))
+
+
+@pytest.mark.parametrize("nmodes", [3, 4, 5, 6])
+def test_number_moments_multimode_thermal(nmodes):
+    r"""Test the correct values of the photon number means and covariances"""
+    assert nmodes > 1
+    n_samples = 100000
+    n_mean = 3.0
+    A = generate_positive_definite_matrix(nmodes)
+    ls, O = rescale_adjacency_matrix_thermal(A, n_mean)
+    Nmat = O @ np.diag(ls / (1.0 - ls)) @ O.T
+    samples = np.array(generate_thermal_samples(ls, O, n_samples))
+
+    nmean_est = samples.mean(axis=0)
+    cov_est = np.cov(samples.T)
+    expected_cov = np.zeros_like(cov_est)
+    for i in range(nmodes):
+        for j in range(i):
+            expected_cov[i, j] = Nmat[i, i] * Nmat[j, j] + Nmat[i, j] ** 2
+            expected_cov[j, i] = expected_cov[i, j]
+        expected_cov[i, i] = 2 * (Nmat[i, i] ** 2) + Nmat[i, i]
+    ## These moments are obtained using Wick's theorem for a multimode thermal state
+    expected_cov = expected_cov - np.outer(np.diag(Nmat), np.diag(Nmat))
+    # To construct the covariance matrix we need to subtract the projector onto the mean
+    assert np.allclose(nmean_est, np.diag(Nmat), rtol=20 / np.sqrt(n_samples))
+    assert np.allclose(expected_cov, cov_est, rtol=20 / np.sqrt(n_samples))
```

Comment:
How long does this test take to run?

Reply:
21.43 seconds in my 4 core computer, which is nothing against a life of shame for not getting the right sample ;)